test/libsolidity/SemanticTest.cpp

```diff
@@ -142,21 +170,40 @@ TestCase::TestResult SemanticTest::runTest(ostream& _stream, string const& _line
 	if (_compileViaYul)
 		AnsiColorized(_stream, _formatted, {BOLD, CYAN}) << _linePrefix << "Running via Yul:" << endl;
 
-	for (auto& test: m_tests)
+	for (TestFunctionCall& test: m_tests)
 		test.reset();
 
 	map<string, solidity::test::Address> libraries;
 
 	bool constructed = false;
 
-	for (auto& test: m_tests)
+	// Iterate through the test calls and set the previous call.
+	TestFunctionCall* previousCall{nullptr};
+	for (TestFunctionCall& test: m_tests)
+	{
+		test.setPreviousCall(previousCall);
```

Comment:
I'm not sure about these. Do we really need such mutating functions?

Reply:
At least for the isoltest event support I need the information of the previous call. I use this to basically go up the call chain to find the creator of events, and ultimately find out whether some events where not consumed. I think this could be also useful for some other TestHooks in the future.

Reply:
I know, but isn't there a solution that does not require modifying the test function call? Can you somehow pass the previous call as a parameter?

Reply:
Also shouldn't events use the `endTestCase` hook?

Reply:
> I know, but isn't there a solution that does not require modifying the test function call? Can you somehow pass the previous call as a parameter?

Thats true. Hmm, maybe we can move this into the hook too. So at first we visit all calls and call an "initialise" function for each call within the hook. So the hook could just create whatever needed, so in the event case it will just note in what order the calls where done. 

This would be a good improvement. Maybe there is also no real need for such "initialise" function, maybe `beforeFunctionCall` could be used somehow. I will look into that.

> Also shouldn't events use the `endTestCase` hook?

Depends. Right now I check for the events after each call that may produced events and check whether, between event producing calls, all events where consumed. That means I check this basically after each call. It could also be done at the end of the test-case. 

However, I would prefer to have this check done after each call, because it may be possible that a call is overwriting the logs of another call.

Reply:
At least right now I think having such "initialise" function would be the most generic thing. What I basically mean is to add a function e.g. `void initFunctionCall(TestFunctionCall const& _call)` to `TestHook`.  Before we call `beginTestCase`, we just do
```cpp
for (TestFunctionCall& test: m_tests)
    for (std::unique_ptr<TestHook> const& hook: m_testHooks)
        hook->initFunctionCall(test);
```
With this the hook can implement whatever the hook needs. 

Regarding `test.setTestHooks(&m_testHooks);`  we could probably just provide the hooks as a parameter to `TestFunctionCall`.

Reply:
Ok, it looks like that this is not really needed. At least for the event support PR. I could imagine that it may be useful to have such `initFunctionCall` mechanism here, but at least for now this is not really needed. @mijovic do you need sich thing for your PR?

However, I will remove the `previousCall` and `TestFunctionCall::setTestHooks(..)`. If someone sees that having something like `initFunctionCall` in `TestHook` maybe useful, I could add it.

Reply:
I think It will not be needed in my gas costs PR
Sounds good to remove them now as it's not needed now, and if there is such need in other PRs, we can easily add it

Reply:
Sorry, I meant `afterFunctionCall` instead of `endTestCase`.